cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -424,6 +429,510 @@ class self_comparator {
 };
 
 }  // namespace lexicographic
+
+namespace hash {
+class row_hasher;
+}
+
+namespace equality {
+
+template <typename Nullate>
+class device_row_comparator {
+  friend class self_comparator;
+
+ public:
+  /**
+   * @brief Checks whether the row at `lhs_index` in the `lhs` table is equal to the row at
+   * `rhs_index` in the `rhs` table.
+   *
+   * @param lhs_index The index of row in the `lhs` table to examine
+   * @param rhs_index The index of the row in the `rhs` table to examine
+   * @return `true` if row from the `lhs` table is equal to the row in the `rhs` table
+   */
+  __device__ bool operator()(size_type const lhs_index, size_type const rhs_index) const noexcept
+  {
+    auto equal_elements = [=](column_device_view l, column_device_view r) {
+      return cudf::type_dispatcher(
+        l.type(), element_comparator{nulls, l, r, nulls_are_equal}, lhs_index, rhs_index);
```

Comment:
Yes. Using this
```c++
void check_shape_compatibility(table_view const& lhs, table_view const& rhs)
{
  std::function<void(column_view const&, column_view const&)> check_column =
    [&](column_view const& l, column_view const& r) {
      CUDF_EXPECTS(l.type().id() == r.type().id(),
                   "Cannot compare tables with different column types");
      CUDF_EXPECTS(l.num_children() == r.num_children(), "Mismatched number of children");
      for (size_type i = 0; i < l.num_children(); ++i) {
        check_column(l.child(i), r.child(i));
      }
    };

  CUDF_EXPECTS(lhs.num_columns() == rhs.num_columns(),
               "Cannot compare tables with different number of columns");
  for (size_type i = 0; i < lhs.num_columns(); ++i) {
    check_column(lhs.column(i), rhs.column(i));
  }
}
```
But we'll add it when we add support for 2 tables. The idea is that such checks should live in the comparator owner because we have access to a `table_view`. The `device_row_comparator` types are not publicly accessible.

Reply:
Okay, this is still single-table. That's what I was forgetting. I'll resolve this.